### PR TITLE
Handle packages inside another node modules package when auto importing

### LINKED ISSUE #37542: Import fix creates incorrect URL for symbols from preact/hooks
**TypeScript Version:**  3.8.3

**Search Terms:**
import preact/hooks preact/hooks/src

**Code**

```tsx
import {h} from 'preact';

function App(): void {
  const value = useMemo(() => 'Hello');
  return <div>{state}</div>;
}
```

Run missing import fix on `useMemo`.

**Expected behavior:**

The quick fix should add the import line

```ts
import {useMemo} from 'preact/hooks';
```

**Actual behavior:**

The quick fix actually adds the incorrect import line

```ts
import {useMemo} from 'preact/hooks/src';
```

This will cause the source files from Preact to be bundled instead of the transpiled files, which will cause various kinds of breakage (see https://github.com/preactjs/preact/issues/1757 or https://github.com/preactjs/preact/issues/1373).

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #37542

### CODE REVIEW

File: tests/cases/fourslash/importFixesWithPackageJsonInSideAnotherPackage.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+////    "jsx": "react",
+////    "jsxFactory": "h"
+////  }
+////}
+
+// @Filename: /project/app.tsx
+////const state = useMemo(() => 'Hello', []);
+
+// @Filename: /project/component.tsx
+////import { useEffect } from "preact/hooks";
```
Comment (ForsakenHarmony) [inline tests/cases/fourslash/importFixesWithPackageJsonInSideAnotherPackage.ts]: this line seems kinda useless without all the context from the repro, should it be removed or does it still serve a purpose?

Comment (sheetalkamat) [reply tests/cases/fourslash/importFixesWithPackageJsonInSideAnotherPackage.ts]: Its used to include preact/hooks/src/index.d.ts to be included in the program

File: src/compiler/moduleSpecifiers.ts [hunk 2]
```diff
@@ -319,33 +319,30 @@ namespace ts.moduleSpecifiers {
...
+                // If the module could be imported by a directory name, use that directory's name
+                const { moduleFileToTry, packageRootPath } = tryDirectoryWithPackageJson(packageRootIndex);
+                if (packageRootPath) {
+                    moduleSpecifier = packageRootPath;
+                    break;
+                }
+                if (!moduleFileNameForExtensionless) moduleFileNameForExtensionless = moduleFileToTry;
+
+                // try with next level of directory
+                packageRootIndex = moduleFileName.indexOf(directorySeparator, packageRootIndex + 1);
```
Comment (andrewbranch) [inline src/compiler/moduleSpecifiers.ts]: Just to make sure I’m understanding this correctly, the idea is:

1. we’ve located the `moduleFileName`, say, `/project/node_modules/preact/hooks/src/index.js`
2. we look at the structure of the path (as before) to call `/project/node_modules/preact` the “package root”
3. we see if `moduleFileName` is accessible (via a package.json main field) from `/project/node_modules/preact`
4. if not, go to step 3 but with one more directory appended to the end (in the first iteration of this case, `/project/node_modules/preact/hooks`)

As contrasted to the previous behavior, where we’d only try the package root before giving up and using the full `moduleFileName` (optionally with `/index.js` taken off the end).

**Comment by marvinhagemeister** [comment]
Thank you so much for fixing this 👍

File: tests/cases/fourslash/importFixesWithPackageJsonInSideAnotherPackage.ts (hunk 1 above)
Comment (ForsakenHarmony) [reply tests/cases/fourslash/importFixesWithPackageJsonInSideAnotherPackage.ts]: ah, right

**Pull request merged**